tests/kernel/semaphore/semaphore/src/main.c

```diff
@@ -960,37 +835,45 @@ void test_sem_multiple_threads_wait(void)
  */
 void test_sem_measure_timeouts(void)
 {
-	int32_t ret_value;
-	uint32_t start_ticks, end_ticks;
+	int64_t start_ticks, end_ticks, diff_ticks;
 
 	k_sem_reset(&simple_sem);
 
 	/* with timeout of 1 sec */
 	start_ticks = k_uptime_get();
 
-	ret_value = k_sem_take(&simple_sem, K_SECONDS(1));
+	expect_k_sem_take_nomsg(&simple_sem, K_SECONDS(1), -EAGAIN);
 
 	end_ticks = k_uptime_get();
 
-	zassert_true(ret_value == -EAGAIN, "k_sem_take failed");
+	diff_ticks = end_ticks - start_ticks;
 
-	zassert_true((end_ticks - start_ticks >= SEC2MS(1)),
-		     "time missmatch - expected %d, got %d",
-		     SEC2MS(1), end_ticks - start_ticks);
+	zassert_true((diff_ticks >= SEC2MS(1)),
+		     "k_sem_take returned too early: %lld < %d",
+		     diff_ticks, SEC2MS(1));
 
-	/* with 0 as the timeout */
+	/* This subtest could fail spuriously if we happened to run the below right as
+	 * a tick occurred. Unfortunately, we cannot actually fix this, because on some
+	 * emulated platforms time does not advance while running the cpu, so
+	 * if we spin and wait for a tick boundary we'll spin forever.
+	 * The best we can do is hope that k_busy_wait finishes just after a tick boundary.
+	 */
+	k_busy_wait(1);
```

Comment:
@andyross - you're probably not the best person for this, but you probably know who is.

I'm somewhat surprised that this hasn't been causing any long-tail failures, although it is a fairly narrow timing window.

Reply:
Actually I did the timer stuff too, so I hate to say it but I kinda am.

And yes: this kind of race-vs-real-time behavior is endemic in the test suite, and not really treatable in a rigorous way.  Systems vary widely across CPU speeds and tick rates.

The most common way to deal with tests where you need reproducible behavior is to start the CPU-bound portion of the test aligned to a tick boundary.  Timer expirations are (or at least should be) arranged by the drivers to be on regular tick intervals, so waiting on any kernel timeout expiration ("`k_usleep(1);`" is the most common idiom) is going to start you as close to the beginning of a tick as the kernel could manage.

Even then there are platforms where the tick rate is so high that getting out of the timer interrupt and context switch and past the setup takes longer than a tick though, so YMMV.

Reply:
Should I replace the `k_busy_wait(1)` call here with `k_usleep(1)` instead?

I originally was directly doing:

```c
static int64_t wait_for_tick_boundary(void) {
  int64_t curr_tick;
  int64_t start_tick = k_uptime_get();
  
  do  {
      curr_tick = k_uptime_get();
  } while (curr_tick == start_tick);
  return curr_tick;
}
```

for the "run ASAP after a tick boundary" portion, but ran into the amusement that the posix emulator doesn't advance time while cpu-bound. (I was debating dropping in a call to advance time one tick every, say, 1000 calls to `hwm_get_time`, but that's _decidedly_ out of scope for this PR.)

*****

> Even then there are platforms where the tick rate is so high that getting out of the timer interrupt and context switch and past the setup takes longer than a tick though, so YMMV.

Yep, obvious example being anything trying to use a cpu-frequency counter for their timer.

Really, this test should probably be rewritten to check what it's _actually_ looking for - namely if this call can yield. But I'm lothe to make test content changes and test format changes in the same PR. (This PR already has too much of that...)

Reply:
That loop is basically what k_busy_wait() does by default FWIW, though some platforms have an optimized version.   But yes, it looks like a usleep tick alignment would be what you want here.  High tick rate systems with tickless enabled can usually do better by polling (because they don't have to take an interrupt), but emulated systems with real time clocks like qemu tend to misbehave if you try that (seemingly because the host kernel seems them as CPU-bound and is more likely to schedule them in and out with a larger quantum).